quota/quota.go

```diff
@@ -0,0 +1,180 @@
+	// special meaning such as user.* or system.* or snapd.*, but are otherwise
+	// generically choosable by a user. This name will correspond to the name of
+	// the systemd slice underlying the quota group.
+	Name string
+
+	// SubGroups is the set of sub-groups that are subject to this quota.
+	// Sub-groups have their own limits, subject to the requirement that the
+	// highest quota for a sub-group is that of the parent group.
+	SubGroups []*Group `json:"sub-groups,omitempty"`
+
+	// MemoryLimit is the limit of memory available to the processes in the
+	// group where if the total used memory of all the processes exceeds the
+	// limit, oom-killer is invoked which will start killing processes. The
+	// specific behavior of which processes are killed is subject to the
+	// ExhaustionBehavior.
+	MemoryLimit quantity.Size `json:"memory-limit,omitempty"`
+
+	// MemoryExhaustionBehavior is the behavior that is implemented to decide
+	// what processes to kill in the quota group when the memory limits are
+	// exhausted.
+	MemoryExhaustionBehavior MemoryExhaustionBehavior `json:"memory-exhaustion-behavior,omitempty"`
+
+	// parentGroup is the the parent group that this group is a child of. It is
+	// here mainly for coding convenience to easily construct the full chain
+	// back to a root group to generate the full slice unit name of the quota
+	// group. If it is nil, then this is a "root" quota group.
+	parentGroup *Group
+}
+
+// MemoryExhaustionBehavior is the behavior determining what the system should
+// do when memory available to a quota group is exhausted.
+type MemoryExhaustionBehavior int
+
+const (
+	// DefaultOOMKiller is the default behavior to invoke to choose which
+	// processes when the memory quota resource limit is reached, and it
+	// consists of letting the Linux kernel's oom-killer run and decide, which
+	// is available on all cgroups version systems, but is not deterministic as
+	// to which processes are killed.
+	DefaultOOMKiller MemoryExhaustionBehavior = iota
+)
+
+// SliceFileName returns the name of the slice file that should be used for this
+// quota group. This name will include all of the group's parents in the name.
+// For example, a group named "bar" that is a child of the "foo" group will have
+// a systemd slice name as "foo-bar.slice". Note that the slice name may differ
+// from the snapd friendly group name, mainly in the case that the group is a
+// sub group.
+func (grp *Group) SliceFileName() string {
+	if grp.parentGroup == nil {
+		// root group name, then the slice unit is just "<name>.slice"
+		return fmt.Sprintf("%s.slice", grp.Name)
+	}
+
+	// otherwise we need to track back to get all of the parent elements
+	grpNames := []string{}
+	parentGrp := grp.parentGroup
+	for parentGrp != nil {
+		grpNames = append([]string{parentGrp.Name}, grpNames...)
+		parentGrp = parentGrp.parentGroup
+	}
+
+	buf := &bytes.Buffer{}
+	for _, parentGrpName := range grpNames {
+		fmt.Fprintf(buf, "%s-", parentGrpName)
```

Comment:
for consistency with our other names we might allow '-' in quota group names which means this function needs to do appropriate quoting like we do for mounts?

Reply:
yes I was hoping to avoid doing the escaping of "-" for simplicity's sake, but it's certainly doable to allow "-" and escape it

Reply:
this is done now